src/cargo/ops/cargo_new.rs

```diff
@@ -862,3 +881,58 @@ mod tests {
 
     Ok(())
 }
+
+fn update_manifest_with_inherited_workspace_package_keys(
+    opts: &MkOptions<'_>,
+    manifest: &mut toml_edit::Document,
+    workspace_package_keys: &toml_edit::Table,
+) {
+    if workspace_package_keys.is_empty() {
+        return;
+    }
+
+    let remove_package_key = |key: &str, manifest: &mut toml_edit::Document| {
+        let package = manifest["package"]
+            .as_table_mut()
+            .expect("package is a table");
+        package.remove(key);
+    };
+
+    let set_workspace_inherited_package_key = |key: &str, manifest: &mut toml_edit::Document| {
+        let package = manifest["package"]
+            .as_table_mut()
+            .expect("package is a table");
+        let mut table = toml_edit::Table::new();
+        table.set_dotted(true);
+        table["workspace"] = toml_edit::value(true);
+        package.insert(key, toml_edit::Item::Table(table));
+    };
+
+    let remove_and_inherit_package_key = |key: &str, manifest: &mut toml_edit::Document| {
+        remove_package_key(key, manifest);
+        set_workspace_inherited_package_key(key, manifest);
+    };
+
+    // Try inherit the edition from the workspace if it is not specified.
+    if opts.edition.is_none() && workspace_package_keys.contains_key("edition") {
+        remove_and_inherit_package_key("edition", manifest);
+    }
+
+    // Try inherit the version from the workspace.
+    if workspace_package_keys.contains_key("version") {
+        remove_and_inherit_package_key("version", manifest);
+    }
+
+    // Try inherit the publish from the workspace if it is not specified.
+    if opts.registry.is_none() && workspace_package_keys.contains_key("publish") {
+        remove_and_inherit_package_key("publish", manifest);
+    }
+
+    // Inherit other keys from the workspace.
+    for (key, _) in workspace_package_keys
+        .iter()
+        .filter(|(key, _)| *key != "edition" && *key != "version" && *key != "publish")
+    {
+        set_workspace_inherited_package_key(key, manifest);
+    }
```

Comment:
Would it work for us to unconditonally remove the package key so we can just have a single loop?

Also, I assume we remove, rather than overwrite, so inherited keys go at the end?  Thats the style I go with but we should probably explicitly state in the PR description stylistic choices like that

Reply:
The downside to not fully looping is we are not fully preserving the order of `[workspace.package]`

Reply:
We need to consider the case where opts are passed. It has a higher priority than inheritance.

Reply:
That isn't the case for `version`.

For `edition` and `publish`, we could just short-circuit the loop (`if key == "edition" && opts.edition.is_some() { continue; }`_).  That will also make  the skipping for user-provided values more obvious (somehow I had missed it in my first pass and had to re-read multiple times before I picked it up).

If nothing else, when looking at the tests, it seems weird that for almost all the fields the order of `[workspace.package]` is preserved but not these.

Reply:
Updated. Sounds good. 
Thanks for your review!  💚 💙 💜 💛 ❤️